File: Chap5-MVCAdvanced/5-18CustomModelsDragAndDrop/personmodel.cpp
#include "personmodel.h"
#include <QMimeData>

PersonModel::PersonModel(QObject *parent) : QAbstractListModel(parent)
{
    persons << "Donald Bush" << "Xi Jing Tao" <<
               "Vladmir Golbathev" <<
               "Emmanuel Mitterand" << "Jacob Mandela";
}

int PersonModel::rowCount(const QModelIndex &parent) const
{
    Q_UNUSED(parent);
    return  persons.size();
}


QVariant PersonModel::data(const QModelIndex &index, int role) const
{
    if (!index.isValid())
        return QVariant();

    if (index.row() < 0 || index.row() >= persons.size())
        return QVariant();
    if (role == Qt::DisplayRole || role == Qt::EditRole)
        return persons.at(index.row());

    return  QVariant();
}

bool PersonModel::setData(const QModelIndex &index, const QVariant &value, int role)
{
    Q_UNUSED(role);

    if(index.isValid()){
        persons.replace(index.row(),value.toString());
        emit dataChanged(index,index);
        return true;
    }
    return  false;
}

QVariant PersonModel::headerData(int section, Qt::Orientation orientation, int role) const
{

    if (role != Qt::DisplayRole)
        return QVariant();

    if(orientation == Qt::Horizontal){
        if(section == 0){
            return "Names";
        }else{
            return  QVariant();
        }
    }

    return QVariant();

}

bool PersonModel::insertRows(int row, int count, const QModelIndex &parent)
{
    Q_UNUSED(parent);
    beginInsertRows(QModelIndex(), row, row+count-1);

    for (int i = 0; i < count; ++i) {
        persons.insert(row, "");
    }
    endInsertRows();
    return true;
}

bool PersonModel::removeRows(int row, int count, const QModelIndex &parent)
{
    Q_UNUSED(parent);
    beginRemoveRows(QModelIndex(), row, row+count-1);

    for (int i = 0; i < count; ++i) {
        persons.removeAt(row);
    }
    endRemoveRows();
    return true;
}

QStringList PersonModel::mimeTypes() const
{
    return QAbstractListModel::mimeTypes() << "plain/text";
}

QMimeData *PersonModel::mimeData(const QModelIndexList &indexes) const
{

    QStringList list;
    for( const QModelIndex & index : indexes){
        list << index.data().toString();
    }
    QMimeData * mimeData = QAbstractListModel::mimeData(indexes);
    mimeData->setText(list.join(","));

    return mimeData;
}

bool PersonModel::dropMimeData(const QMimeData *data, Qt::DropAction action,
                               int row, int column, const QModelIndex &parent)
{

    if(data->hasText()){

        if(parent.isValid()){
            //Ovewrite
            setData(parent,data->text(),Qt::DisplayRole);
        }else{
            //Add new data
            insertRows(rowCount(),1,QModelIndex());
            setData(index(rowCount()-1,0,QModelIndex()),data->text());

        }

        return true;
    }
    return QAbstractListModel::dropMimeData(data,action,row,column,parent);
}

Qt::ItemFlags PersonModel::flags(const QModelIndex &index) const
{
    if(!index.isValid()){
        return Qt::ItemIsEnabled | Qt::ItemIsDropEnabled;
    }else{
        return QAbstractListModel::flags(index) | Qt::ItemIsEditable | Qt::ItemIsDragEnabled |
                Qt::ItemIsDropEnabled;
    }
}

File: Chap5-ModelViewArchitectureAdvanced/5-18CustomModelsDragAndDrop/personmodel.py
from PySide6.QtCore import (
    QAbstractListModel, QModelIndex, Qt, QMimeData
)

class PersonModel(QAbstractListModel):
    def __init__(self, parent=None):
        super().__init__(parent)
        self._persons = [
            "Donald Bush", 
            "Xi Jing Tao", 
            "Vladmir Golbathev", 
            "Emmanuel Mitterand", 
            "Jacob Mandela"
        ]

    def rowCount(self, parent=QModelIndex()):
        """Return the number of rows in the model."""
        return len(self._persons)

    def data(self, index, role=Qt.DisplayRole):
        """Provide data for the given index and role."""
        if not index.isValid():
            return None

        if index.row() < 0 or index.row() >= len(self._persons):
            return None

        if role in (Qt.DisplayRole, Qt.EditRole):
            return self._persons[index.row()]

        return None

    def setData(self, index, value, role=Qt.EditRole):
        """Update data for a given index."""
        if role == Qt.EditRole and index.isValid():
            self._persons[index.row()] = str(value)
            self.dataChanged.emit(index, index)
            return True
        return False

    def headerData(self, section, orientation, role=Qt.DisplayRole):
        """Provide header data."""
        if role == Qt.DisplayRole and orientation == Qt.Horizontal:
            return "Names" if section == 0 else None
        return None

    def insertRows(self, row, count=1, parent=QModelIndex()):
        """Insert rows into the model."""
        self.beginInsertRows(parent, row, row + count - 1)
        for _ in range(count):
            self._persons.insert(row, "")
        self.endInsertRows()
        return True

    def removeRows(self, row, count=1, parent=QModelIndex()):
        """Remove rows from the model."""
        self.beginRemoveRows(parent, row, row + count - 1)
        del self._persons[row:row+count]
        self.endRemoveRows()
        return True

    def mimeTypes(self):
        """Define supported MIME types."""
        return super().mimeTypes() + ["plain/text"]

    def mimeData(self, indexes):
        """Create MIME data for drag operations."""
        mime_data = super().mimeData(indexes)
        text_data = ",".join(str(index.data()) for index in indexes)
        mime_data.setText(text_data)
        return mime_data

    def dropMimeData(self, data, action, row, column, parent):
        """Handle drop operations."""
        if isinstance(data, dict) and "text" in data:
            # For QML integration
            text = data["text"]
        elif hasattr(data, "hasText") and data.hasText():
            # For Qt Widgets
            text = data.text()
        else:
            return False

        if parent.isValid():
            # Overwrite existing item
            self.setData(parent, text)
        else:
            # Add new item at the end
            self.insertRows(self.rowCount())
            self.setData(self.index(self.rowCount() - 1), text)

        return True

    def flags(self, index):
        """Define item flags for drag and drop."""
        if not index.isValid():
            return Qt.ItemIsEnabled | Qt.ItemIsDropEnabled

        return (super().flags(index) | 
                Qt.ItemIsEditable | 
                Qt.ItemIsDragEnabled | 
                Qt.ItemIsDropEnabled)
                
    def roleNames(self):
        """Define role names for QML."""
        return {Qt.DisplayRole: b"display"}